[Comb] Narrow Arithmetic and Bitwise Ops - Take 2

fyquah opened the pull request:
This is a second attempt in implementing #1424 , fixing some memory access bugs.

There are two memory access issues:
1. What @lattner mentioned about uses in https://github.com/llvm/circt/commit/2660c0b3144be11370839ebfaad8c0c80b335f82#r53799782 . I fixed this by dumping `operation->getUsers` into a vector .
2. Calling `rewriter.replaceOp` ([here](https://github.com/llvm/circt/commit/2660c0b3144be11370839ebfaad8c0c80b335f82#diff-6f80524cbed3f3dee4e82a5f97ae9ade89be7f2132704918ec60d91889bfc751R159)) erases the op under the hood and invalidates that piece of memory.  This was fine when we are replacing the Op being canonicalized, but reasons i'm not sure of, erasing users of the op is buggy.  valgrind confirms that the GreedyPatternRewriter has a dangling reference (see log below), but i can't convince myself why. I fixed this by using `rewriter.replaceAllUsesOfOp`, which does not erase the op's memory. This ends up achieving the desired result, as DCE clears out the unused op.

To help with review, this PR comprises of 3 commits (and more as we iterate through this in review):
1. https://github.com/llvm/circt/commit/9f25b9ce6a521561192c6c77793fb9b1d0d3eb27 Readds the broken patch (ie revert the revert)
2. https://github.com/llvm/circt/commit/02313deb99a2b828777d90e7474a45ad4967a2df Implements the fixes. I move to useing `getUsers` instead of `getUse`, as it returns an iterator over  `Operation*` directly (which is what i care about).
3. https://github.com/llvm/circt/commit/529ab72292ff5f0abf532cf2a8bef11370f3be65 Implements some review comments in discussed in https://github.com/llvm/circt/commit/2660c0b3144be11370839ebfaad8c0c80b335f82 , namely changing `template<typename T>`  into `Operation*`, and `typename F` into `std::function<Value(ArrayRef(Value))>` 

To validate that this works:

1. I ran the Comb cannonicalization test in this feature 10k times
2. Running the HEAD of this PR with valgrind using `./bin/circt-opt --canonicalize ../test/Dialect/Comb/canonicalization.mlir` gets a clean result.

*Verbose Valgrind output follows*

The following is the output from valgrind when ran against the broken commit. This illustrates the problem where `getUses()` memory is being wipped as it is being rewritten.

```
fyq14@linux:~/dev/circt/build$ valgrind ./bin/circt-opt --canonicalize  ../test/Dialect/Comb/canonicalization.mlir 
==33635== Memcheck, a memory error detector
==33635== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==33635== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==33635== Command: ./bin/circt-opt --canonicalize ../test/Dialect/Comb/canonicalization.mlir
==33635== 
==33635== Invalid read of size 8
==33635==    at 0x66168DC: mlir::detail::IROperandBase::getNextOperandUsingThisValue() (mlir/include/mlir/IR/UseDefLists.h:42)
==33635==    by 0x6758DAC: mlir::ValueUseIterator<mlir::OpOperand>::operator++() (mlir/include/mlir/IR/UseDefLists.h:264)
==33635==    by 0x6752EC3: mlir::Operation::UseIterator::operator++() (mlir/lib/IR/Operation.cpp:1334)
==33635==    by 0x755B75: _Z20narrowOperationWidthIN5circt4comb5MuxOpEZZL18narrowExtractWidthNS1_9ExtractOpEN4mlir15PatternRewriterEENKUlT_E2_clIS2_EEDaS6_EUlN4llvm8ArrayRefINS4_5ValueEEEE_EbS6_NS4_10ValueRangeEbRS5_T0_ (CombFolds.cpp:151)
==33635==    by 0x752632: _ZZL18narrowExtractWidthN5circt4comb9ExtractOpEN4mlir15PatternRewriterEENKUlT_E2_clINS0_5MuxOpEEEDaS4_ (CombFolds.cpp:381)
==33635==    by 0x7526C5: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::MuxOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#4}>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#4}&&) (TypeSwitch.h:116)
==33635==    by 0x749F36: narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter) (CombFolds.cpp:384)
==33635==    by 0x74A0B8: circt::comb::ExtractOp::canonicalize(circt::comb::ExtractOp, mlir::PatternRewriter&) (CombFolds.cpp:410)
==33635==    by 0x785871: mlir::RewritePatternSet& mlir::RewritePatternSet::add<circt::comb::ExtractOp>(mlir::LogicalResult (*)(circt::comb::ExtractOp, mlir::PatternRewriter&))::FnPattern::matchAndRewrite(circt::comb::ExtractOp, mlir::PatternRewriter&) const (PatternMatch.h:978)
==33635==    by 0x78BEB3: mlir::detail::OpOrInterfaceRewritePatternBase<circt::comb::ExtractOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (PatternMatch.h:329)
==33635==    by 0x53883AF: mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (mlir/lib/Rewrite/PatternApplicator.cpp:201)
==33635==    by 0x5143BE0: (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:218)
==33635==  Address 0x9f6a340 is 96 bytes inside a block of size 128 free'd
==33635==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33635==    by 0x674C557: mlir::Operation::destroy() (mlir/lib/IR/Operation.cpp:224)
==33635==    by 0x674D114: llvm::ilist_traits<mlir::Operation>::deleteNode(mlir::Operation*) (mlir/lib/IR/Operation.cpp:432)
==33635==    by 0x66273E4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>) (llvm/include/llvm/ADT/ilist.h:268)
==33635==    by 0x6756EE4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(mlir::Operation*) (llvm/include/llvm/ADT/ilist.h:272)
==33635==    by 0x674D36F: mlir::Operation::erase() (mlir/lib/IR/Operation.cpp:481)
==33635==    by 0x676CF25: mlir::RewriterBase::replaceOp(mlir::Operation*, mlir::ValueRange) (mlir/lib/IR/PatternMatch.cpp:230)
==33635==    by 0x755AF8: _Z20narrowOperationWidthIN5circt4comb5MuxOpEZZL18narrowExtractWidthNS1_9ExtractOpEN4mlir15PatternRewriterEENKUlT_E2_clIS2_EEDaS6_EUlN4llvm8ArrayRefINS4_5ValueEEEE_EbS6_NS4_10ValueRangeEbRS5_T0_ (CombFolds.cpp:159)
==33635==    by 0x752632: _ZZL18narrowExtractWidthN5circt4comb9ExtractOpEN4mlir15PatternRewriterEENKUlT_E2_clINS0_5MuxOpEEEDaS4_ (CombFolds.cpp:381)
==33635==    by 0x7526C5: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::MuxOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#4}>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#4}&&) (TypeSwitch.h:116)
==33635==    by 0x749F36: narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter) (CombFolds.cpp:384)
==33635==    by 0x74A0B8: circt::comb::ExtractOp::canonicalize(circt::comb::ExtractOp, mlir::PatternRewriter&) (CombFolds.cpp:410)
==33635==  Block was alloc'd at
==33635==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33635==    by 0x674B8E6: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, unsigned int) (mlir/lib/IR/Operation.cpp:136)
==33635==    by 0x674BE56: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, mlir::RegionRange) (mlir/lib/IR/Operation.cpp:94)
==33635==    by 0x674BD63: mlir::Operation::create(mlir::OperationState const&) (mlir/lib/IR/Operation.cpp:83)
==33635==    by 0x662F093: mlir::OpBuilder::createOperation(mlir::OperationState const&) (mlir/lib/IR/Builders.cpp:400)
==33635==    by 0x78F0589: (anonymous namespace)::OperationParser::parseCustomOperation(llvm::ArrayRef<std::tuple<llvm::StringRef, unsigned int, llvm::SMLoc> >) (mlir/lib/Parser/Parser.cpp:1922)
==33635==    by 0x78EC806: (anonymous namespace)::OperationParser::parseOperation() (mlir/lib/Parser/Parser.cpp:801)
==33635==    by 0x78F9111: (anonymous namespace)::OperationParser::parseBlockBody(mlir::Block*) (mlir/lib/Parser/Parser.cpp:2131)
==33635==    by 0x78F8797: (anonymous namespace)::OperationParser::parseBlock(mlir::Block*&) (mlir/lib/Parser/Parser.cpp:2099)
==33635==    by 0x78F830B: (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:2057)
==33635==    by 0x78F763E: (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:1996)
==33635==    by 0x78F46DB: (anonymous namespace)::CustomOpAsmParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::OperandType>, llvm::ArrayRef<mlir::Type>, bool) (mlir/lib/Parser/Parser.cpp:1645)
==33635==
<snip>
```

The following diff and valgrind log shows a fix for the `getUses` getting invalidated as we replace uses, and the valgrind log the follows. Note that the greedy pattern driver still contain dangling references.

```diff
diff --git a/lib/Dialect/Comb/CombFolds.cpp b/lib/Dialect/Comb/CombFolds.cpp
index 7b6bbc48..931259e1 100644
--- a/lib/Dialect/Comb/CombFolds.cpp
+++ b/lib/Dialect/Comb/CombFolds.cpp
@@ -148,8 +148,11 @@ static bool narrowOperationWidth(Op narrowingCandidate, ValueRange inputs,
 
   // Replace all the use-site's extract operation with the newly minted narrowed
   // version.
-  for (auto &use : uses) {
-    auto extractUse = cast<ExtractOp>(use.getOwner());
+  SmallVector<Operation*> users(
+      narrowingCandidate.getOperation()->getUsers().begin(),
+      narrowingCandidate.getOperation()->getUsers().end());
+  for (auto &user : users) {
+    auto extractUse = cast<ExtractOp>(user);
     auto oldLowBit = extractUse.lowBit();
 
     assert(oldLowBit >= lowestBitRequired &&
```


```
fyq14@linux:~/dev/circt/build$ valgrind ./bin/circt-opt --canonicalize  ../test/Dialect/Comb/canonicalization.mlir 
==33936== Memcheck, a memory error detector
==33936== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==33936== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==33936== Command: ./bin/circt-opt --canonicalize ../test/Dialect/Comb/canonicalization.mlir
==33936== 
==33936== Invalid read of size 4
==33936==    at 0x63B95B0: mlir::Operation::getResults() (mlir/include/mlir/IR/Operation.h:293)
==33936==    by 0x63B8B04: mlir::Operation::getOpResults() (mlir/include/mlir/IR/Operation.h:297)
==33936==    by 0x63B8A24: mlir::Operation::use_empty() (mlir/include/mlir/IR/Operation.h:578)
==33936==    by 0x63B80C4: mlir::isOpTriviallyDead(mlir::Operation*) (mlir/lib/Interfaces/SideEffectInterfaces.cpp:33)
==33936==    by 0x5143937: (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:180)
==33936==    by 0x51434E3: mlir::applyPatternsAndFoldGreedily(llvm::MutableArrayRef<mlir::Region>, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:256)
==33936==    by 0x4F29D1B: (anonymous namespace)::Canonicalizer::runOnOperation() (mlir/lib/Transforms/Canonicalizer.cpp:48)
==33936==    by 0x541E557: mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (mlir/lib/Pass/Pass.cpp:385)
==33936==    by 0x541EB44: mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::unique_ptr<mlir::Pass, std::default_delete<mlir::Pass> >*, mlir::Pass> >, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (mlir/lib/Pass/Pass.cpp:445)
==33936==    by 0x542022D: mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) (mlir/lib/Pass/Pass.cpp:689)
==33936==    by 0x5420117: mlir::PassManager::run(mlir::Operation*) (mlir/lib/Pass/Pass.cpp:669)
==33936==    by 0x4860F88: performActions(llvm::raw_ostream&, bool, bool, llvm::SourceMgr&, mlir::MLIRContext*, mlir::PassPipelineCLParser const&) (mlir/lib/Support/MlirOptMain.cpp:84)
==33936==  Address 0x9fb7aa4 is 52 bytes inside a block of size 128 free'd
==33936==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674C557: mlir::Operation::destroy() (mlir/lib/IR/Operation.cpp:224)
==33936==    by 0x674D114: llvm::ilist_traits<mlir::Operation>::deleteNode(mlir::Operation*) (mlir/lib/IR/Operation.cpp:432)
==33936==    by 0x66273E4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>) (llvm/include/llvm/ADT/ilist.h:268)
==33936==    by 0x6756EE4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(mlir::Operation*) (llvm/include/llvm/ADT/ilist.h:272)
==33936==    by 0x674D36F: mlir::Operation::erase() (mlir/lib/IR/Operation.cpp:481)
==33936==    by 0x676CF25: mlir::RewriterBase::replaceOp(mlir::Operation*, mlir::ValueRange) (mlir/lib/IR/PatternMatch.cpp:230)
==33936==    by 0x758E5D: bool narrowOperationWidth<circt::comb::AddOp, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}) (CombFolds.cpp:162)
==33936==    by 0x757AB3: bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&) (CombFolds.cpp:184)
==33936==    by 0x755143: _ZZL18narrowExtractWidthN5circt4comb9ExtractOpEN4mlir15PatternRewriterEENKUlT_E_clINS0_5AddOpEEEDaS4_ (CombFolds.cpp:356)
==33936==    by 0x7551D5: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::AddOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&) (TypeSwitch.h:116)
==33936==    by 0x7522AC: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::Operation*, bool>, mlir::Operation*>::Case<circt::comb::AddOp, circt::comb::MulOp, , narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&&) (TypeSwitch.h:41)
==33936==  Block was alloc'd at
==33936==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674B8E6: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, unsigned int) (mlir/lib/IR/Operation.cpp:136)
==33936==    by 0x674BE56: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, mlir::RegionRange) (mlir/lib/IR/Operation.cpp:94)
==33936==    by 0x674BD63: mlir::Operation::create(mlir::OperationState const&) (mlir/lib/IR/Operation.cpp:83)
==33936==    by 0x662F093: mlir::OpBuilder::createOperation(mlir::OperationState const&) (mlir/lib/IR/Builders.cpp:400)
==33936==    by 0x78F0589: (anonymous namespace)::OperationParser::parseCustomOperation(llvm::ArrayRef<std::tuple<llvm::StringRef, unsigned int, llvm::SMLoc> >) (mlir/lib/Parser/Parser.cpp:1922)
==33936==    by 0x78EC806: (anonymous namespace)::OperationParser::parseOperation() (mlir/lib/Parser/Parser.cpp:801)
==33936==    by 0x78F9111: (anonymous namespace)::OperationParser::parseBlockBody(mlir::Block*) (mlir/lib/Parser/Parser.cpp:2131)
==33936==    by 0x78F8797: (anonymous namespace)::OperationParser::parseBlock(mlir::Block*&) (mlir/lib/Parser/Parser.cpp:2099)
==33936==    by 0x78F830B: (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:2057)
==33936==    by 0x78F763E: (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:1996)
==33936==    by 0x78F46DB: (anonymous namespace)::CustomOpAsmParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::OperandType>, llvm::ArrayRef<mlir::Type>, bool) (mlir/lib/Parser/Parser.cpp:1645)
==33936== 
```

lattner commented:
> Calling rewriter.replaceOp (here) erases the op under the hood and invalidates that piece of memory. This was fine when we are replacing the Op being canonicalized, but reasons i'm not sure of, erasing users of the op is buggy. valgrind confirms that the GreedyPatternRewriter has a dangling reference (see log below), but i can't convince myself why. I fixed this by using rewriter.replaceAllUsesOfOp, which does not erase the op's memory. This ends up achieving the desired result, as DCE clears out the unused op.

My memory on this is fuzzy, and it is probably worth asking @River707 or others on the MLIR channel, but looking in PatternMatch.cpp, I see this:

```
/// This method replaces the results of the operation with the specified list of
/// values. The number of provided values must match the number of results of
/// the operation.
void RewriterBase::replaceOp(Operation *op, ValueRange newValues) {
  // Notify the rewriter subclass that we're about to replace this root.
  notifyRootReplaced(op);
```

From the looks of that, it looks like "op" is only allowed to be the root in question.  If this is true, we should probably add some assertions to this, as I suspect others will hit the same problem, and tracking down this memory invalidation bug sounds like a real beast of an issue.

In any case, I think your fix is quite reasonable.

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,128 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+
+  // It is not necessary to call rewriter.replaceOp(narrowingCandidate,
+  // narrowedOperation) here, since all the use-site now refers to
+  // `narrowedOperation` instead. narrowingCandidate will be eliminated by DCE.
```
You can move this comment up to where this happens :-)

fyquah replied:
I reworded this to something this captures both the non-root case and root-case better =)

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,128 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+  // version.
+  SmallVector<Operation *, 16> users(usersIterator.begin(),
+                                     usersIterator.end());
+  for (auto *user : users) {
```
Does it work to do:

```
while (!narrowingCandidate->use_empty()) {
  auto *user = *narrowingCandidate->user_begin();
```

If this doesn't work for some reason, please sink the definition of `usesIterator` down here and pass it directly into the smallvector ctor

mikeurbach replied:
There is also `llvm::make_early_inc_range`, which I've used for this kind of thing: https://github.com/llvm/circt/blob/main/lib/Conversion/HWToLLHD/HWToLLHD.cpp#L315

fyquah replied:
I didn't know about `llvm::make_early_inc_range` -- thanks showing it!

lattner reviewed (commented):
Thank you again for pushing on this!

mikeurbach reviewed (commented):
> not just use RAUW

Yep, the [docs](https://mlir.llvm.org/docs/PatternRewriter/#restrictions) mention "All IR mutations, including creation, _must_ be performed by the given `PatternRewriter`". Directly using an API like RAUW is not allowed. However, if you need to use something like that, I think you can use the PatternRewriter's `updateRootInPlace` to mark a _different_ root operation temporarily and do any IR mutation you need in a callback. For example, I do that [here](https://github.com/llvm/circt/blob/02fb43f665dfa768c0acc4d6a90e1e734318c6bd/lib/Conversion/HWToLLHD/HWToLLHD.cpp#L317).

River707 commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+narrowOperationWidth(ExtractOp outerExtractOp, Operation *narrowingCandidate,
+                     ValueRange inputs, bool narrowTrailingBits,
+                     PatternRewriter &rewriter,
+                     std::function<Value(ArrayRef<Value>)> createOp) {
```
Please use function_ref instead of std::function.

https://llvm.org/docs/ProgrammersManual.html#the-function-ref-class-template

River707 commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+    auto loc = extractUse.getLoc();
+
+    if (narrowedWidth == extractUse.getType().getWidth()) {
+      return Value(narrowedOperation);
```
Can you instead explicitly define the result type of this lambda (i.e. `-> Value`).

River707 commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+  rewriter.replaceOp(outerExtractOp,
+                     getNarrowedExtractReplacement(outerExtractOp));
+
+  // However, for other use sites, calling rewriter.replaceOp or
```
You can use replaceOp on any op, it isn't required to be the root operation. If any bugs arise, that would be a bug in the pattern driver/documentation.

fyquah replied:
I thought so, but I changed  `rewriter.replaceOp` to `RAUW` after running into some very occasional segfaults. The valgrind output is as follows. From what I can make of it, it looks like there are dangling references to none root ops in the Greedypatternrewriter worklist.

(This is the same valgrind output that follows is same as those in the PR description above).

```
fyq14@linux:~/dev/circt/build$ valgrind ./bin/circt-opt --canonicalize  ../test/Dialect/Comb/canonicalization.mlir 
==33936== Memcheck, a memory error detector
==33936== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==33936== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==33936== Command: ./bin/circt-opt --canonicalize ../test/Dialect/Comb/canonicalization.mlir
==33936== 
==33936== Invalid read of size 4
==33936==    at 0x63B95B0: mlir::Operation::getResults() (mlir/include/mlir/IR/Operation.h:293)
==33936==    by 0x63B8B04: mlir::Operation::getOpResults() (mlir/include/mlir/IR/Operation.h:297)
==33936==    by 0x63B8A24: mlir::Operation::use_empty() (mlir/include/mlir/IR/Operation.h:578)
==33936==    by 0x63B80C4: mlir::isOpTriviallyDead(mlir::Operation*) (mlir/lib/Interfaces/SideEffectInterfaces.cpp:33)
==33936==    by 0x5143937: (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:180)
==33936==    by 0x51434E3: mlir::applyPatternsAndFoldGreedily(llvm::MutableArrayRef<mlir::Region>, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:256)
==33936==    by 0x4F29D1B: (anonymous namespace)::Canonicalizer::runOnOperation() (mlir/lib/Transforms/Canonicalizer.cpp:48)
==33936==    by 0x541E557: mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (mlir/lib/Pass/Pass.cpp:385)
==33936==    by 0x541EB44: mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::unique_ptr<mlir::Pass, std::default_delete<mlir::Pass> >*, mlir::Pass> >, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (mlir/lib/Pass/Pass.cpp:445)
==33936==    by 0x542022D: mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) (mlir/lib/Pass/Pass.cpp:689)
==33936==    by 0x5420117: mlir::PassManager::run(mlir::Operation*) (mlir/lib/Pass/Pass.cpp:669)
==33936==    by 0x4860F88: performActions(llvm::raw_ostream&, bool, bool, llvm::SourceMgr&, mlir::MLIRContext*, mlir::PassPipelineCLParser const&) (mlir/lib/Support/MlirOptMain.cpp:84)
==33936==  Address 0x9fb7aa4 is 52 bytes inside a block of size 128 free'd
==33936==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674C557: mlir::Operation::destroy() (mlir/lib/IR/Operation.cpp:224)
==33936==    by 0x674D114: llvm::ilist_traits<mlir::Operation>::deleteNode(mlir::Operation*) (mlir/lib/IR/Operation.cpp:432)
==33936==    by 0x66273E4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>) (llvm/include/llvm/ADT/ilist.h:268)
==33936==    by 0x6756EE4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(mlir::Operation*) (llvm/include/llvm/ADT/ilist.h:272)
==33936==    by 0x674D36F: mlir::Operation::erase() (mlir/lib/IR/Operation.cpp:481)
==33936==    by 0x676CF25: mlir::RewriterBase::replaceOp(mlir::Operation*, mlir::ValueRange) (mlir/lib/IR/PatternMatch.cpp:230)
==33936==    by 0x758E5D: bool narrowOperationWidth<circt::comb::AddOp, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}) (CombFolds.cpp:162)
==33936==    by 0x757AB3: bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&) (CombFolds.cpp:184)
==33936==    by 0x755143: _ZZL18narrowExtractWidthN5circt4comb9ExtractOpEN4mlir15PatternRewriterEENKUlT_E_clINS0_5AddOpEEEDaS4_ (CombFolds.cpp:356)
==33936==    by 0x7551D5: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::AddOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&) (TypeSwitch.h:116)
==33936==    by 0x7522AC: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::Operation*, bool>, mlir::Operation*>::Case<circt::comb::AddOp, circt::comb::MulOp, , narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&&) (TypeSwitch.h:41)
==33936==  Block was alloc'd at
==33936==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674B8E6: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, unsigned int) (mlir/lib/IR/Operation.cpp:136)
==33936==    by 0x674BE56: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, mlir::RegionRange) (mlir/lib/IR/Operation.cpp:94)
==33936==    by 0x674BD63: mlir::Operation::create(mlir::OperationState const&) (mlir/lib/IR/Operation.cpp:83)
==33936==    by 0x662F093: mlir::OpBuilder::createOperation(mlir::OperationState const&) (mlir/lib/IR/Builders.cpp:400)
==33936==    by 0x78F0589: (anonymous namespace)::OperationParser::parseCustomOperation(llvm::ArrayRef<std::tuple<llvm::StringRef, unsigned int, llvm::SMLoc> >) (mlir/lib/Parser/Parser.cpp:1922)
==33936==    by 0x78EC806: (anonymous namespace)::OperationParser::parseOperation() (mlir/lib/Parser/Parser.cpp:801)
==33936==    by 0x78F9111: (anonymous namespace)::OperationParser::parseBlockBody(mlir::Block*) (mlir/lib/Parser/Parser.cpp:2131)
==33936==    by 0x78F8797: (anonymous namespace)::OperationParser::parseBlock(mlir::Block*&) (mlir/lib/Parser/Parser.cpp:2099)
==33936==    by 0x78F830B: (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:2057)
==33936==    by 0x78F763E: (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:1996)
==33936==    by 0x78F46DB: (anonymous namespace)::CustomOpAsmParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::OperandType>, llvm::ArrayRef<mlir::Type>, bool) (mlir/lib/Parser/Parser.cpp:1645)
==33936== 

```

River707 replied:
Can you file a bug with a repro and assign it to me? This should be made to work.

River707 commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -219,20 +355,70 @@ static LogicalResult extractConcatToConcatExtract(ExtractOp op,
+        return narrowOperationWidth(outerExtractOp, innerOp, innerOp.inputs(),
+                                    /* narrowTrailingBits= */ false, rewriter);
+      })
+      .Case<SubOp>([&](auto innerOp) {
```
Not sure if you have a preference, but for the single type cases I'm pretty sure you could also do:

 `.Case([&](SubOp innerOp) ...`

fyquah commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,128 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+      auto narrowedExtract = rewriter.create<ExtractOp>(
+          loc, extractUse.getType(), narrowedOperation, newLowBit);
+      extractUse.replaceAllUsesWith(narrowedExtract.getOperation());
+    }
```
I tried doing that, but that results in some ocassional segfault and valgrind complaining about bad memory accesses. Under the hood, it looks like `GredyPatternRewritter::replaceOp` replaces all uses, then calls `rewriter.eraseOp`, so this would amount to the same effect.

For now it only looks like it's safe to call this on the rootOp (which uses replaceOp anyway).

But maybe I'm simply doing something memory unsafe here,  let's see what @River707 says below.

fyquah commented:
I've addressed some review here, but I have not tested any of this with valgrind yet (I don't physically have a machine that can do that right now), but will do so in a few days time :)

River707 commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+  //
+  // Calling narrowingCandidate->getUsers will not include outerExtractOp, since
+  // rewriter.replaceOp will have removed it from the iterator.
+  for (Operation *user :
```
Are there situations where narrowing candidate is used more than once by the same operation? In those cases, the user would appear more than once in the `getUsers` list.

fyquah replied:
Yes, they could appear multiple times. I think the current code will do the correct thing, but will do some unnecessary duplicated work.

FYI, addressing this issue using the following patch doesn't get rid of the seg faults:

```diff
-  for (Operation *user :
-       llvm::make_early_inc_range(narrowingCandidate->getUsers())) {
+  while (!narrowingCandidate->use_empty()) {
+    Operation *user = *narrowingCandidate->user_begin();
     auto extractUser = cast<ExtractOp>(user);
     rewriter.updateRootInPlace(extractUser, [&]() {
-      extractUser.replaceAllUsesWith(
-          getNarrowedExtractReplacement(extractUser));
+        rewriter.replaceOp(extractUser, getNarrowedExtractReplacement(extractUser));
     });
   }
```

fyquah replied:
actually -- i'm mistaken, in this specific instance, it will never appear multiple times. All uses of narrowingCandidate will be an `comb::ExtractOp`, which has only 1 operand.

fyquah commented:
@River707 @lattner  On trying to create a reprod, i found the true source of the bug. This little diff is self-explainatory :) 

```diff
 // usage sites of outerExtractOp to reference the newly created narrowed
 // operation.
 static bool narrowExtractWidth(ExtractOp outerExtractOp,
-                               PatternRewriter rewriter) {
+                               PatternRewriter &rewriter) {
   auto *innerArg = outerExtractOp.input().getDefiningOp();
```

I created a commit (69655864ee38167016506e9dae2eb1eb43dc3ba5) that changes RAUW -> `rewriter.replaceOp`, and removes threading `outerExtractOp` down.

lattner commented:
Oof, good catch.  We should disable copy construction of the pattern rewriter!!

lattner reviewed (approved):
This looks great to me, thank you for tracking this down!

lattner commented:
Patch sent upstream here: https://reviews.llvm.org/D107232

seldridge commented:
Ack. We started to see segfaults after this landed this morning. I blamelessly reverted in 6a1e44a784c5e10c9e907a95b2d2760fbcd012e9. (And thanks for your hard work on this @fyquah! I'm sorry it's proving tedious!)

Here's the segfault that I'm currently seeing:

```
0  firtool                  0x0000000109515b3d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 61
1  firtool                  0x00000001095160eb PrintStackTraceSignalHandler(void*) + 27
2  firtool                  0x0000000109513d4b llvm::sys::RunSignalHandlers() + 123
3  firtool                  0x0000000109517da8 SignalHandler(int) + 232
4  libsystem_platform.dylib 0x00007fff2040ed7d _sigtramp + 29
5  libsystem_platform.dylib 0xfffffffffffff000 _sigtramp + 18446603339975033504
6  firtool                  0x0000000109801224 llvm::ilist_base<true>::insertBeforeImpl(llvm::ilist_node_base<true>&, llvm::ilist_node_base<true>&) + 68
7  firtool                  0x0000000109a262e3 void llvm::ilist_base<true>::insertBefore<llvm::ilist_node_impl<llvm::ilist_detail::node_options<mlir::Operation, true, false, void> > >(llvm::ilist_node_impl<llvm::ilist_detail::node_options<mlir::Operation, true, false, void> >&, llvm::ilist_node_impl<llvm::ilist_detail::node_options<mlir::Operation, true, false, void> >&) + 35
8  firtool                  0x0000000109a262a6 llvm::simple_ilist<mlir::Operation>::insert(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>, mlir::Operation&) + 54
9  firtool                  0x0000000109a26257 llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::insert(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>, mlir::Operation*) + 71
10 firtool                  0x00000001098b141c mlir::OpBuilder::insert(mlir::Operation*) + 76
11 firtool                  0x00000001098b1730 mlir::OpBuilder::createOperation(mlir::OperationState const&) + 48
12 firtool                  0x0000000109d73486 circt::comb::ExtractOp mlir::OpBuilder::create<circt::comb::ExtractOp, mlir::IntegerType, mlir::Value&, unsigned int&>(mlir::Location, mlir::IntegerType&&, mlir::Value&, unsigned int&) + 454
13 firtool                  0x0000000109d71861 narrowOperationWidth(mlir::Operation*, mlir::ValueRange, bool, mlir::PatternRewriter&, llvm::function_ref<mlir::Value (llvm::ArrayRef<mlir::Value>)>)::$_23::operator()(circt::comb::ExtractOp) const + 305
14 firtool                  0x0000000109d70e4a narrowOperationWidth(mlir::Operation*, mlir::ValueRange, bool, mlir::PatternRewriter&, llvm::function_ref<mlir::Value (llvm::ArrayRef<mlir::Value>)>) + 1050
15 firtool                  0x0000000109d74ede bool narrowOperationWidth<circt::comb::OrOp>(circt::comb::OrOp, mlir::ValueRange, bool, mlir::PatternRewriter&) + 174
16 firtool                  0x0000000109d74e1f auto narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20::operator()<circt::comb::OrOp>(circt::comb::OrOp) const + 95
17 firtool                  0x0000000109d74c96 llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::OrOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20&>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20&) + 134
18 firtool                  0x0000000109d745e5 llvm::TypeSwitch<mlir::Operation*, bool>& llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::Operation*, bool>, mlir::Operation*>::Case<circt::comb::OrOp, circt::comb::XorOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20&>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20&) + 37
19 firtool                  0x0000000109d705f1 llvm::TypeSwitch<mlir::Operation*, bool>& llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::Operation*, bool>, mlir::Operation*>::Case<circt::comb::AndOp, circt::comb::OrOp, circt::comb::XorOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&)::$_20&&) + 49
20 firtool                  0x0000000109d6289b narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter&) + 139
21 firtool                  0x0000000109d6203e circt::comb::ExtractOp::canonicalize(circt::comb::ExtractOp, mlir::PatternRewriter&) + 334
```

From what I can tell, it looks like the rewriter may be pointing at a bad op when it tries to insert a new one.  The following looks suspicious::

```
(lldb) frame select 8
frame #8: 0x0000000100a2d861 firtool`narrowOperationWidth(this=0x000070000ed00b30, extractUse=ExtractOp @ 0x000070000ed00970)>)::$_23::operator()(circt::comb::ExtractOp) const at CombFolds.cpp:167:38
   164 	    }
   165 	
   166 	    uint32_t newLowBit = oldLowBit - lowestBitRequired;
-> 167 	    Value narrowedExtract = rewriter.create<ExtractOp>(
   168 	        loc, extractUse.getType(), narrowedOperation, newLowBit);
   169 	    return narrowedExtract;
   170 	  };
(lldb) print rewriter.insertPoint->dump()
<<UNKNOWN SSA VALUE>> = comb.extract %82 from 1 : (i3) -> i2
```

seldridge commented:
(Also, I'm seeing what I can do to reduce the test case for you.)

seldridge commented:
Here's a reduced test case:

```mlir
module {
  hw.module @Foo(%a: i8) -> (%out: i1) {
    %0 = comb.extract %a from 4 : (i8) -> i4
    %1 = comb.extract %a from 0 : (i8) -> i4
    %2 = comb.or %0, %1 : i4
    %3 = comb.extract %2 from 2 : (i4) -> i2
    %4 = comb.extract %2 from 0 : (i4) -> i2
    %5 = comb.or %3, %4 : i2
    %6 = comb.extract %5 from 0 : (i2) -> i1
    hw.output %6 : i1
  }
}
```

This will crash if compiled with: 

```
circt-opt -canonicalize='top-down=true region-simplify=true' Foo.mlir
```

lattner commented:
nice @seldridge
